src/main.ts

```diff
@@ -54,14 +60,48 @@ export class Minifier {
     var pos = file.getLineAndCharacterOfPosition(start);
     var fullMessage = `${fileName}:${pos.line + 1}:${pos.character + 1}: ${message}`;
     this.errors.push(fullMessage);
-    if (minifierOptions.failFast) {
+    if (this.minifierOptions.failFast) {
       throw new Error(fullMessage);
     }
   }
 
+  renameProgram(fileNames: string[], destination?: string) {
+    var host = ts.createCompilerHost(options);
+    var program = ts.createProgram(fileNames, options, host);
+    this.typeChecker = program.getTypeChecker();
+
+    program.getSourceFiles()
+        .filter((sf) => !sf.fileName.match(/\.d\.ts$/))
+        .forEach((f) => {
+          var renamedTSCode = this.visit(f);
+          var fileName = this.getOutputPath(f.fileName, destination);
+          fsx.mkdirsSync(path.dirname(fileName));
+          fs.writeFileSync(fileName, renamedTSCode);
+        });
+  }
+
+  getOutputPath(filePath: string, destination?: string): string {
+    // convert everything to absolute paths
+    var destination = path.resolve(process.cwd(), destination || '.');
+    var filePath = path.resolve(process.cwd(), filePath);
+
+    // no base path, flatten file structure and output to destination
+    if (!this.minifierOptions.basePath) {
+      return path.join(destination, path.basename(filePath));
+    }
+
+    if (!path.isAbsolute(this.minifierOptions.basePath)) {
+      this.minifierOptions.basePath = path.resolve(process.cwd(), this.minifierOptions.basePath);
+    }
+
+    // given a base path, preserve file directory structure
+    var subFilePath = filePath.replace(this.minifierOptions.basePath, '');
```

Comment:
what if the basePath occurs somewhere within the filePath, but not at the beginning?